File: Object Pool/Ruby/ThreadPool.rb
require_relative "Thread.rb"

class ThreadPool
    def initialize
        @threads = []
        @counter = 2

        for i in 1..2 do
            thread = TThread.new(i)
            @threads << thread
        end
    end

    def get
        thread = nil

        if @threads.length > 0
            thread = @threads.shift
        else
            @counter += 1

            thread = TThread.new(@counter)
        end

        return thread
    end

    def release(thread)
        @threads << thread
    end
end

File: Object Pool/Javascript/ThreadPool.js
import Thread from "./Thread.js";

class ThreadPool {
  constructor() {
    this.threads = [];
    this.counter = 2;

    for (let index = 0; index < this.counter; index++) {
      this.threads.push(new Thread(index + 1));
    }
  }

  get() {
    let thread = null;

    if (this.threads.length > 0) {
      thread = this.threads.shift();
    } else {
      this.counter++;
      thread = new Thread(this.counter);
    }

    return thread;
  }

  release(thread) {
    this.threads.push(thread);
  }
}

export default ThreadPool;